ckan/logic/action/__init__.py

```diff
@@ -73,3 +75,76 @@ def prettify(field_name: str):
         else:
             summary[_(prettify(key))] = error[0]
     return summary
+
+
+# (canada fork only): handle all errors in resource actions
+# TODO: upstream contrib??
+def resource_validation_errors(
+        error_dict: ErrorDict,
+        action: str,
+        pkg_dict: Dict[str, Any],
+        resource_index: int = -1) -> Tuple[Dict[str, Any], str]:
+    """
+    Checks through the error_dict to find all errors in
+    the Dataset and its Resources.
+    """
+    new_error_dict = dict(error_dict)
+    # define out own error summaries so we can have
+    # a more customized structure to the ErrorDict
+    error_summaries = []
+    try:
+        if action == 'delete':
+            # special case for deleting as there is no index
+            # for a non-existent resource in the pkg_dict.
+            current_res_error_dict = False
+        else:
+            current_res_error_dict = cast("list[ErrorDict]", error_dict['resources'])[resource_index]
+        if current_res_error_dict:
+            # if there are errors for the current resource
+            # let's raise them to the user first.
+            new_error_dict = current_res_error_dict
```

Comment:
```suggestion
            new_error_dict = dict(current_res_error_dict)
```
because new_error_dict may be modified below